### PR TITLE
Make region flag optional

### PR DESCRIPTION
**Pull request by prateekgogia**
Made the region flag to be optional, and fetch the region from metadata service in ec2.

## Testing
- Test on EC2 instance without region flag set and discovered from metadata
```
/tmp/aws-encryption-provider --listen=/tmp/kmsplugin/socket.sock --key=arn:aws:kms:us-west-2:<account>:key/<ID>
{"level":"info","timestamp":"2022-03-10T04:04:17.655Z","caller":"server/main.go:68","message":"creating kms server","health-port":":8080","healthz-path":"/healthz","livez-path":"/livez","region":"","listen-address":"/tmp/kmsplugin/socket.sock","kms-endpoint":"","qps-limit":0,"burst-limit":0}
{"level":"info","timestamp":"2022-03-10T04:04:17.655Z","caller":"plugin/plugin.go:333","message":"registering the kms plugin with grpc server"}
{"level":"info","timestamp":"2022-03-10T04:04:17.655Z","caller":"server/main.go:101","message":"Healthchecks server started","port":":8080"}
{"level":"info","timestamp":"2022-03-10T04:04:17.655Z","caller":"server/main.go:102","message":"Plugin server started","port":"/tmp/kmsplugin/socket.sock"}
```

- Test with region flag set
```
 /tmp/aws-encryption-provider --listen=/tmp/kmsplugin/socket.sock --key=arn:aws:kms:us-west-2:<account>:key/<ID> --region=us-west-2
{"level":"info","timestamp":"2022-03-10T04:27:37.825Z","caller":"server/main.go:68","message":"creating kms server","health-port":":8080","healthz-path":"/healthz","livez-path":"/livez","region":"us-west-2","listen-address":"/tmp/kmsplugin/socket.sock","kms-endpoint":"","qps-limit":0,"burst-limit":0}
{"level":"info","timestamp":"2022-03-10T04:27:37.826Z","caller":"plugin/plugin.go:333","message":"registering the kms plugin with grpc server"}
{"level":"info","timestamp":"2022-03-10T04:27:37.826Z","caller":"server/main.go:101","message":"Healthchecks server started","port":":8080"}
{"level":"info","timestamp":"2022-03-10T04:27:37.826Z","caller":"server/main.go:102","message":"Plugin server started","port":"/tmp/kmsplugin/socket.sock"}
```

- Test when region flag is not set and metadata is not available, it should error out
```
./bin/aws-encryption-provider --listen=/tmp/kmsplugin/socket.sock --key=arn:aws:kms:us-west-2:<account>:key/<ID>
{"level":"info","timestamp":"2022-03-09T21:57:23.150-0600","caller":"server/main.go:64","message":"creating kms server","health-port":":8080","healthz-path":"/healthz","livez-path":"/livez","region":"","listen-address":"/tmp/kmsplugin/socket.sock","kms-endpoint":"","qps-limit":0,"burst-limit":0}
{"level":"fatal","timestamp":"2022-03-09T21:57:29.454-0600","caller":"server/main.go:76","message":"Failed to create new KMS service","error":"failed to call the metadata server's region API, EC2MetadataRequestError: failed to get EC2 instance identity document\ncaused by: RequestError: send request failed\ncaused by: Get \"http://169.254.169.254/latest/dynamic/instance-identity/document\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)","stacktrace":"main.main\n\t/Users/pgogia/workspace/golang/src/github.com/kubernetes-sigs/aws-encryption-provider/cmd/server/main.go:76\nruntime.main\n\t/usr/local/Cellar/go/1.17.6/libexec/src/runtime/proc.go:255"}
```

### CODE REVIEW

File: pkg/cloud/cloud.go [hunk 1]
```diff
@@ -28,17 +29,22 @@ type AWSKMS struct {
 }
 
 func New(region, kmsEndpoint string, qps, burst int) (*AWSKMS, error) {
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new session: %w", err)
+	}
+	if region == "" {
+		region, err = ec2metadata.New(sess).Region()
```
Comment (micahhausler) [inline pkg/cloud/cloud.go:37]: Is this required? Doesn't the SDK do this for you if no region is specified?

File: pkg/cloud/cloud.go [hunk 2]
```diff
@@ -28,17 +29,22 @@ type AWSKMS struct {
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to call the metadata server's region API, %v", err)
+		}
+	}
+	sess.Config.Region = aws.String(region)
```
[the remark this answers is not in the document]
Comment (prateekgogia) [reply pkg/cloud/cloud.go:42]: yeah, true we can skip setting it here. Sending an update.

File: pkg/cloud/cloud.go (hunk 1 above)
Comment (micahhausler) [reply pkg/cloud/cloud.go:37]: Yea, the region field won't be set on the session config, but try initiating a service like KMS and just listing keys without a region set.  
```go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/kms"
)

func main() {
	svc := kms.New(session.Must(session.New()))
	result, err := svc.ListKeys(&kms.ListKeysInput{})
	if err != nil {
		panic(err)
	}
	data, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		panic(err)
	}
	fmt.Println(string(data))
}
```

Comment (prateekgogia) [reply pkg/cloud/cloud.go:37]: I tried to list with this code and it fails to list saying `MissingRegion: could not find region configuration`.

I also tried to run this [sample client](https://github.com/kubernetes-sigs/aws-encryption-provider/blob/master/cmd/client/main.go) it connects with the server, but fails to encrypt / decrypt if region is not set
```
2022/03/10 18:26:56 Failed to encrypt: rpc error: code = Unknown desc = failed to encrypt MissingRegion: could not find region configuration
```

Comment (micahhausler) [reply pkg/cloud/cloud.go:37]: ok cool, TIL! I figured the region was automatically set

**Review by micahhausler — APPROVED** [review]
/lgtm
/approve

**Pull request merged**